### ISSUE TITLE
How can I do single-cls training?

### ISSUE THREAD

**Issue by Rich2020** [ultralytics/ultralytics#1490]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

In YOLOv5, we could use the `--single-cls` option to do only object detection.
How is this done in YOLOv8? I tried using `single_cls`, however in my logs, I see loads of warnings saying `... ignoring corrupt image/label: Label class 5 exceeds dataset class count 1. Possible class labels are 0-0`. 

I have tried having my config list all the classes:

```
names:
  0: aaa
  1: bbb
  2: ccc
  ...
  ...

```

And I have also tried listing just 1 class:

```
names:
  0: aaa
```

In both cases, I get the same warning.

**Comment by glenn-jocher** [comment]
@Rich2020 thanks for asking. This feature may not be functional yet. I'll add a TODO to track this.

@Laughing-q seems like single-cls is not working correctly. I'll take a look.

**Comment by Laughing-q** [comment]
@Rich2020 @glenn-jocher works to me with the latest version.
```bash
yolo detect train model=yolov8n.yaml data=runs/coco128.yaml imgsz=640 single_cls
```
here's the training samples:
![train_batch0](https://user-images.githubusercontent.com/61612323/226173263-3db55f1b-4a4a-41d4-b995-505b08df98a0.jpg)

**Comment by Rich2020** [comment]
@glenn-jocher @Laughing-q Thank you both -- I have just checked and can confirm that @Laughing-q is correct that everything is being classed as 0. However, the warning messages probably need to be checked, just in case something is awry.

**Comment by glenn-jocher** [comment]
@Rich2020 I don't see any warnings when training with single_cls, everything works correctly for me. 

```
from ultralytics import YOLO

model=YOLO('yolov8n.pt')
model.train(data='coco128.yaml', single_cls=True, imgsz=160, epochs=1)

Ultralytics YOLOv8.0.54 🚀 Python-3.10.6 torch-1.13.1 CPU
yolo/engine/trainer: task=detect, mode=train, model=yolov8n.pt, data=coco128.yaml, epochs=1, patience=50, batch=16, imgsz=160, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=None, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=True, image_weights=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=/Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40
                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.Conv                  [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.Conv                  [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.C2f                   [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.Conv                  [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.C2f                   [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.Conv                  [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.C2f                   [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.Conv                  [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.C2f                   [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.SPPF                  [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.Concat                [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.C2f                   [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.Concat                [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.C2f                   [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.Conv                  [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.Concat                [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.C2f                   [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.Conv                  [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.Concat                [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.C2f                   [384, 256, 1]                 
 22        [15, 18, 21]  1    897664  ultralytics.nn.modules.Detect                [80, [64, 128, 256]]          
Model summary: 225 layers, 3157200 parameters, 3157184 gradients, 8.9 GFLOPs
Transferred 355/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40', view at http://localhost:6006/
optimizer: SGD(lr=0.01) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias
train: Scanning /Users/glennjocher/PycharmProjects/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
val: Scanning /Users/glennjocher/PycharmProjects/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
Plotting labels to /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40/labels.jpg... 
Image sizes 160 train, 160 val
Using 0 dataloader workers
Logging results to /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40
Starting training for 1 epochs...
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
        1/1         0G       1.93      4.733      1.541        206        160: 100%|██████████| 8/8 [00:04<00:00,  1.79it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.45it/s]
                   all        128        929      0.559      0.267      0.308       0.18
1 epochs completed in 0.003 hours.
Optimizer stripped from /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40/weights/last.pt, 6.5MB
Optimizer stripped from /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40/weights/best.pt, 6.5MB
Validating /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40/weights/best.pt...
Ultralytics YOLOv8.0.54 🚀 Python-3.10.6 torch-1.13.1 CPU
Model summary (fused): 168 layers, 3151904 parameters, 0 gradients, 8.7 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:02<00:00,  1.55it/s]
                   all        128        929      0.563      0.268      0.309       0.18
                person        128        929      0.563      0.268      0.309       0.18
Speed: 0.0ms preprocess, 12.8ms inference, 0.0ms loss, 0.2ms postprocess per image
Results saved to /Users/glennjocher/PycharmProjects/ultralytics/runs/detect/train40
```

Training also works correctly with CLI:

```
yolo train single_cls
```

We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by Rich2020** [comment]
@glenn-jocher How odd. I am definitely getting those warnings. I will try deleting the train.cache and try again.
For reference, I used kicked off training as such:

`yolo detect train data=config.yml single_cls model=yolov8l.pt epochs=1000 imgsz=640 device=\'0,1,2,3\'`

**Comment by Rich2020** [comment]
I can confirm that deleting the labels/train.cache and labels/test.cache files resolved the issue -- warnings are gone.
Thank you @glenn-jocher and @Laughing-q for your support!

**Comment by glenn-jocher** [comment]
@Rich2020 got it! Thanks for the letting us know. I think I'll increment the cache version to force a recache of all datasets for all users in 8.0.55 since it seems like older caches are causing issues. Done in https://github.com/ultralytics/ultralytics/pull/1475/commits/99f9b37289b89081787a6c2b05d1d21b95cda2e0

**Comment by seu-su-ai-csw** [comment]
When the dateset has only one categroy, the cls loss should be zero or not?
I used the same dateset to do detect and the dateset has only one categroy, why in yolov5 the cls loss was zero but in yolov8 it wasn't?

**Comment by glenn-jocher** [comment]
@seu-su-ai-csw cls loss should never be zero. YOLOv5 has a different output structure with an extra objectness output that YOLOv8 has eliminated.

**Comment by dankernel** [comment]
I have successfully completed single-class detection training following the method you suggested. 
However, the last layer (Detect) arguments still show `80: [80, [64, 128, 256]]`. I think it should be `[1, [64, 128, 256]]`, with 1 instead of 80.

Additionally, even when single-class is set to True, it appears to generate cache for all classes when reading label data and creating cache. In the verify_image_label function of ultralytics/yolo/data/utils.py

I modified the verify_image_label function in `ultralytics/yolo/data/utils.py` to filter a specific class and forcibly change it to 0 as follows:

```
112                 # class #N -> class #0
113                 lb = lb[lb[:, 0] == 10]
114                 lb[:, 0] = 0
```

**Comment by pderrenger** [comment]
@dankernel-ai-csw That's great to hear that you completed the single-class detection training successfully! Regarding your observations on the last layer arguments and the generation of cache for all classes, I'd recommend creating a new issue on the GitHub repository with these findings to discuss these potential improvements further. It would be beneficial to the community to further explore these suggestions regarding the last layer arguments and the generation of cache. Thank you for your valuable input!

**Comment by glenn-jocher** [comment]
@pderrenger-csw It's great that you've successfully completed your training. Regarding the Detect layer showing `[80, [64, 128, 256]]`, please ensure that your `data.yaml` file is correctly configured for a single class with `nc: 1`. This should reflect in the Detect layer as `[1, [64, 128, 256]]`.

Your approach to modifying `verify_image_label` function works, but adjusting the source code might not be necessary or the best practice for most users. Also, when `single_cls` is set to True, the cache generation should ideally be aware of this and only consider the single class you are training on, rather than all classes. This sounds like something we may need to look into improving.

If you believe that modifying the `verify_image_label` function is essential for your training process and beneficial for others, feel free to contribute back to the community by suggesting these changes in the GitHub repository. We welcome community contributions and discussions that help improve the models and training process. Thanks for sharing your process and insights!
